pkg/extprom/instrument_server.go

```diff
@@ -0,0 +1,75 @@
+package extprom
+
+import (
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var (
+	requestDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "http_request_duration_seconds",
+			Help: "Tracks the latencies for HTTP requests.",
+		},
+		[]string{"code", "handler", "method"},
+	)
+
+	requestSize = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: "http_request_size_bytes",
+			Help: "Tracks the size of HTTP requests.",
+		},
+		[]string{"code", "handler", "method"},
+	)
+
+	requestsTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "http_requests_total",
+			Help: "Tracks the number of HTTP requests.",
+		}, []string{"code", "handler", "method"},
+	)
+
+	responseSize = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: "http_response_size_bytes",
+			Help: "Tracks the size of HTTP responses.",
+		},
+		[]string{"code", "handler", "method"},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(requestDuration, requestSize, requestsTotal, responseSize)
```

Comment:
generally in thanos there are no init functions used for instrumentation purposes, they are instantiated and registered where used and not global (generally a good practice that we're also pushing everywhere else :) )

Best would be to create this at the very top and pass it down into each function/router that wants to instrument its handler.

Reply:
I have removed `init` and package level vars and introduced new a new type to wrap metrics. I don't think `Instrumentor` is an actual word hence I'm open to suggestions. I used `Register` functions to inject this new behaviour, still not sure if this solution elegant enough 🤔